music/src/core/visualizer.ts

```diff
@@ -456,3 +458,237 @@ export class PianoRollSVGVisualizer extends BaseVisualizer {
+        scrollType: config.scrollType || ScrollType.PAGE,      
+      },
+      div
+    );
+    this.drawnNotes = sequence.notes.length;
+    this.clear();
+    this.redraw();
+  }
+
+  /**
+   * Clears and resets the visualizer object for further redraws from scratch.
+   */
+  protected clear() {
+    this.render.clear();
+  }
+
+  /**
+   * Redraws the entire `noteSequence` in a staff if no `activeNote` is given,
+   * highlighting on and off the appropriate notes otherwise. Should the 
+   * `noteSequence` had changed adding more notes at the end, calling this
+   * method again would complete the redrawing from the very last note it was
+   * drawn, maintaining the active note and the scroll position as they were. 
+   * This is handy for incremental compositions. Given the complexity of 
+   * adaption to a modified score, modifyied notes previously drawn will be
+   * ignored (you can always `clear()` and `redraw()` for a full redraw). 
+   * Please have in mind `mm.Player` does not have this incremental capability
+   * so, once the player had started, it will go on ignoring the changes. 
+   * 
+   * @param activeNote (Optional) If specified, this `Note` will be painted
+   * in the active color and there won't be an actual redrawing, but a 
+   * re-colouring of the involved note heads, accidentals, dots and ties 
+   * (activated and de-activated ones). Otherwise, all musical symbols which 
+   * were not processed yet will be drawn to complete the score.
+   * @param scrollIntoView (Optional) If specified and the active note to be 
+   * highlited is not visualized in the container DIV, the later will be 
+   * scrolled so that the note is viewed in the right place. This can be 
+   * altered by `AdvancedVisualizerConfig.scrollType`.
+   * @returns The x position of the highlighted active note relative to the 
+   * beginning of the SVG, or -1 if there wasn't any given active note. Useful
+   * for automatically advancing the visualization if needed.
+   */
+  public redraw(
+    activeNote?: NoteSequence.INote, 
+    scrollIntoView?: boolean 
+  ) : number {
+    if (this.drawnNotes !== this.noteSequence.notes.length) {
+      this.render.scoreInfo = this.getScoreInfo(this.noteSequence);
+    }
+    const activeNoteInfo = activeNote ? 
+      this.getNoteInfo(activeNote) : undefined;
+    return this.render.redraw(activeNoteInfo, scrollIntoView);
+  }
+
+  private isNoteInInstruments(note: NoteSequence.INote) : boolean {
+    if (note.instrument === undefined || this.instruments.length === 0) {
+      return true; // No instrument information in note means no filtering
+    }
+    else { // Instrument filtering
+      return this.instruments.indexOf(note.instrument) >= 0;
+    }
+  }
+
+  private timeToQuarters(time: number): number {
+    const q = time * this.noteSequence.tempos[0].qpm / 60;
+    return Math.round(q*16)/16; // Current resolution = 1/16 quarter
+  }
+
+  private quantizedStepsToQuarters(steps: number): number {
```

Comment:
you could combine this and the above function into one:
```
q = this.sequenceIsQuantized ? 
   value / this.noteSequence.quantizationInfo.stepsPerQuarter : 
   value * this.noteSequence.tempos[0].qpm / 60;
```

Reply:
Nope, this has a good reason to stay as is:

If you deal just with notes, you are right, because `sequenceIsQuantized` identifies both cases (`sequence.note.startTime` and `sequence.note.quantizedStartStep`), **BUT** (and it is a huge but), protocol buffer definition of `NoteSequence` missed to keep this duality into `sequence.tempos`, `sequence.keySignatures` and `sequence.timeSignatures`because all of them hold an array of elements referenced by `time`but there's not a `quantizedStep`, so if I rely on `sequenceIsQuantized` it could fail because `time` is always meaning seconds, no matter what `sequenceIsQuantized` says.

I double-checked MIDI translation and it always fills time with seconds. That's the reason I intentionally call time or steps version in notes translation and force to time in signatures.

I guess I'm the first one who deals with tempos, key signatures and time signatures and nobody noticed it before. Music is wider than C key and 4/4.

That's re reason of the latter `TODO:` comment you mention.

Reply:
sg, but then i would add these details to that todo (maybe just one of them, so that there's some details about it).